Public clients + CORS docs (#8)
ddubson/docs-tap: app-sso/service-operators/cors.hbs.md

> # Public clients and CORS

Reviewer: **[suggestion]** Either inline or as a "references" section, add:

1. Public / private client definitions as per OAuth 2.1 spec: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-07#section-2.1
2. PKCE spec: https://www.rfc-editor.org/rfc/rfc7636.html

**[suggestion]** Somewhere in the page, for testing purposes, provide a link to [PKCE > Appendix B.  Example for the S256 code_challenge_method](https://www.ietf.org/rfc/rfc7636.html#appendix-B) ; or even the direct values in that document (with a reference): `code_challenge=E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM` and `code_verifier=dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk`

Author: Added suggestions to "References" section.

>**Note** A _public client_ is a client application that does not require credentials to obtain tokens, such as single-page
> apps (SPAs) or mobile devices. Public clients rely on Proof Key for Code Exchange (PKCE) Authorization Code flow extension.

To configure an `AuthServer` and `ClientRegistration`s for use with public clients (single-page apps, mobile devices),

> the following steps are **required**:

Reviewer: **[nit]** Technically, you could still use a `client_secret` in a native (mobile) app - as long as you also use PKCE, so it's not **required**.

Author: I will relax this to "involved" (instead of required)

- [**Allowed HTTP origin(s) must be explicitly specified via `AuthServer.spec.cors` API**](#cors-configuration) - when specified, the
  authorization server will relax the same-origin policy for the specified domain(s), enabling a public client to

>   interact with the designated authorization server.

Reviewer: **[nit]** Technically, this is only for browser-based interactions, not native (mobile app) clients. I don't really have a good way of phrasing it though.

Maybe "this applies to browser-based (single-page) applications"

Author: Sure, will change. Thanks for catching that.

> - [**Client authentication method is set to `none` within `ClientRegistration` resource**](#client-authentication) - public clients are not able
>   to reasonably protect sensitive data such as client secrets, and so client authentication
>   must be turned off.

Reviewer: **[nit]** you could use both PKCE and client_secret ; setting `clientAuthenticationMethod` to `none` forces the Client to use PKCE.

```suggestion
- [**Client authentication method is set to `none` within `ClientRegistration` resource**](#client-authentication) - native applications and browser-based applications are considered public clients
  and should not rely on client credentials. Instead, PKCE should be used. Setting `clientAuthenticationMethod: none`
  ensures client credentials are not used, and makes PKCE mandatory for those clients.
```

## <a id="cors-configuration"></a> CORS configuration

> A public client will be able to interact with an `AuthServer` if the `AuthServer` has the public clients' origin(s)

Reviewer: ```suggestion
A browser-based (public) client will be able to interact with an `AuthServer` if the `AuthServer` has the public clients' origin(s)
```

specified within `AuthServer.spec.cors`.

_Example_: designating specific origins (recommended approach)

```yaml
kind: AuthServer
# ...
spec:
  cors:
    allowOrigins:
    - "https://example.com"        # Specific domain
    - "https://mydept.example.com" # Specific subdomain
    - "https://*.example.com"      # Subdomain wildcard notation
    - "https://*.apps.example.com" # Subdomain wildcard notation
```

> _Example_: designating that all origins are allowed. You must use `allow-unsafe-cors` annotation when defining all 
> origin allowance.

Reviewer: **[suggestion]**

```suggestion
_Example_: designating that all origins are allowed. You must use `allow-unsafe-cors` annotation when defining all 
origin allowance. This AuthServer will send the `Access-Control-Allow-Origin: *` header.
```

> **Caution** Allowing all origins should not be used in production environments.

```yaml
kind: AuthServer
metadata:
  annotations:
    sso.apps.tanzu.vmware.com/allow-unsafe-cors: ""
spec:
  cors:
    allowAllOrigins: true
```

**Requirements for allowed origin designations**

> - An allowed origin **must not** contain `localhost` domain. Use `127.0.0.1` instead.

Reviewer: **[question]** I do not think this is enforced in our code ; this is only about the `redirect_uri` IIRC.

- If this _should_ be enforced in our code, we should have a story for that.
- If it it should not, then we remove this.

Author: Ah gotcha. will remove

> - You **cannot** use both `allowOrigins` and `allowAllOrigins` -- choose one.

Reviewer: **[nit]** wordsmithing

```suggestion
- Only one of `allowOrigins` or `allowAllOrigins` can be set.
```

- When using `allowAllOrigins`, you **must** explicitly set
  annotation `sso.apps.tanzu.vmware.com/allow-unsafe-cors: ""`. This is an acknowledgement that using `allowAllOrigins`
  will be used, but is inherently unsafe.

## <a id="client-authentication"></a>Client authentication

When configuring a `ClientRegistration` for a public client, you must set your client authentication method to
`none` and ensure that your public client supports Authorization Code with PKCE. With PKCE,
the client does not authenticate, but presents a code challenge and subsequent code verifier to establish trust with 
the authorization server.

To set Client Authentication Method to `none`, ensure your `ClientRegistration` resource defines the following:

```yaml
kind: ClientRegistration
# ...
spec:
  clientAuthenticationMethod: none
```

Public clients that support Authorization Code with PKCE flow ensure that:

> - On every OAuth `authorize` request, parameters `code_challenge` and `code_challenge_method` (default: `S256`) are
>   provided.

Reviewer: **[suggestion]**
```suggestion
- On every OAuth `authorize` request, parameters `code_challenge` and `code_challenge_method`. Only `code_challenge_method=S256` is supported.
```
